src/kv_storage.c

```diff
@@ -0,0 +1,494 @@
+    }
+
+    // Check if the current pair is the last one
+    if (iterator->store_index + 2 >= next_store_index) {
+        *has_next = false;
+    } else {
+        *has_next = true;
+    }
+    return ESP_OK;
+}
+
+esp_err_t kv_storage_iterator_next(kv_storage_iterator_t *iterator)
+{
+    bool has_next = false;
+    esp_err_t esp_err = kv_storage_iterator_peek(iterator, &has_next);
+    if (esp_err != ESP_OK) {
+        return esp_err;
+    }
+
+    if (!has_next) {
+        return ESP_ERR_NVS_NOT_FOUND;
+    }
+
+    // Advance the iterator
+    iterator->store_index += 2;
+
+    return ESP_OK;
+}
+
+esp_err_t kv_storage_iterator_get_element(kv_storage_iterator_t *iterator, char *out_key,
+    size_t *out_key_len, void *out_value, size_t *out_value_len)
+{
+    if (iterator->type != NVS_TYPE_BLOB) {
+        ESP_LOGE(TAG, "Calling getter function for binary blob over an iterator of other type.");
+        return ESP_FAIL;
+    }
+
+    // Step 1: Get the key using the store index contained in the iterator
+    char key_entry_name[NVS_KEY_NAME_MAX_SIZE] = { 0 };
+    esp_err_t esp_err = get_entry_name(iterator->store_index, key_entry_name);
+    if (esp_err != ESP_OK) {
+        return ESP_FAIL;
+    }
+    size_t key_len = 0;
+    esp_err = nvs_get_str(iterator->handle, key_entry_name, NULL, &key_len);
+    if (esp_err != ESP_OK) {
+        ESP_LOGE(TAG, "Error getting the key length for %s", key_entry_name);
+        return ESP_FAIL;
+    }
+    // If out_key is null return only the required size for out_key to be stored
+    if (!out_key) {
+        *out_key_len = key_len;
+    } else if (key_len > *out_key_len) {
+        ESP_LOGE(TAG, "Output buffer out_key is insufficient to store the key.");
+        return ESP_ERR_INVALID_SIZE;
+    } else {
+        // Confusing for clang-tidy as second parameter is called 'key'
+        // NOLINTNEXTLINE(readability-suspicious-call-argument)
+        esp_err = nvs_get_str(iterator->handle, key_entry_name, out_key, &key_len);
+        if (esp_err != ESP_OK) {
+            ESP_LOGE(TAG, "Error fetching key from nvs during iteration.");
+            return esp_err;
+        }
+    }
+
+    // Step 2: Get the value using the store index contained in the iterator
+    char value_entry_name[NVS_KEY_NAME_MAX_SIZE] = { 0 };
+    esp_err = get_entry_name(iterator->store_index + 1, value_entry_name);
+    if (esp_err != ESP_OK) {
+        return ESP_FAIL;
+    }
+    size_t value_len = 0;
+    esp_err = nvs_get_blob(iterator->handle, value_entry_name, NULL, &value_len);
+    if (esp_err != ESP_OK) {
+        ESP_LOGE(TAG, "Error fetching value from nvs during erase operation.");
+        return esp_err;
+    }
+    // If out_value is null return only the required size for out_value to be stored
+    if (!out_value) {
+        *out_value_len = value_len;
+        return ESP_OK;
+    }
+    if (value_len > *out_value_len) {
+        ESP_LOGE(TAG, "Output buffer out_value is insufficient to store the value.");
```

Comment:
It would be better to set out_value_len to 0 in case of an error.
```suggestion
        ESP_LOGE(TAG, "Output buffer out_value is insufficient to store the value.");
        *out_value_len = 0;
```

Reply:
Right now the behaviour is the following:
- If `out_value` is null use `out_value_len` as an output to signal the expected size.
- If `out_value` is not null use `out_value_len` as an input representing the size of the `out_value` buffer.

I would not modify `out_value_len` if `out_value` is defined.